distributed/comm/core.py

```diff
@@ -205,14 +206,17 @@ def _raise(error):
     # This starts a thread
     while True:
         try:
-            future = connector.connect(
-                loc, deserialize=deserialize, **(connection_args or {})
-            )
-            comm = await gen.with_timeout(
-                timedelta(seconds=deadline - time()),
-                future,
-                quiet_exceptions=EnvironmentError,
-            )
+            while deadline - time() > 0:
+                future = connector.connect(
+                    loc, deserialize=deserialize, **(connection_args or {})
+                )
+                with ignoring(gen.TimeoutError):
+                    comm = await gen.with_timeout(
+                        timedelta(seconds=1), future, quiet_exceptions=EnvironmentError
```

Comment:
What if the user-specified timeout is quite short (we do this in tests, for example).  

```suggestion
                        timedelta(seconds=min(deadline - time(), 1)), 
                        future, 
                        quiet_exceptions=EnvironmentError
```

Reply:
Ah! Good point.